### PR TITLE
Refactor TrainerModel and NN_opt functions to remove unused code and improve verbosity control

### LINKED ISSUE #2: The `TrainerModel` class should contain only one default model
Including multiple default models as `staticmethod` is confusing. We can provide these models as a subclass of `TrainerModel` and init it with `super()`.

https://github.com/vindolineKis/QCircLearning/blob/25c1983882feefe685bf0a27dec32eebd6593b10/QCircLearning/trainerModel.py#L45

### LINKED ISSUE #5: unit test for : gradients vlues?=0; fitting preformence on a few data.
Write the unit test:
1. check the gradient values ?= 0 while fitting and searching.
2. test the fitting performance of the neural network model on a few data.

### PR DESCRIPTION
**Pull request by cfenglv**

### CODE REVIEW

File: QCircLearning/trainerModel.py [hunk 1]
```diff
@@ -127,18 +112,21 @@ def NN_opt(func, x0, callback=None, **kwargs):
     sample_y = [func(para) for para in sample_x]
     optimal = [sample_x[np.argmin(sample_y)], np.min(sample_y)]
 
-    print("Training with the neural networks")
-    sys.stdout.flush()
+    if verbose:
+        print("Training with the neural networks")
+        sys.stdout.flush()
```
Comment (Canoming) [inline QCircLearning/trainerModel.py:72]: Are these `flush()` necessary?

File: QCircLearning/trainerModel.py [hunk 2]
```diff
@@ -127,18 +112,21 @@ def NN_opt(func, x0, callback=None, **kwargs):
...
 
     for model in nn_models:
-        print(model)
-        sys.stdout.flush()
+        if verbose:
+            print(model)
+            sys.stdout.flush()
```
Comment (Canoming) [inline QCircLearning/trainerModel.py:77]: Maybe we can elaborate more on what is going on for each iteration. For example we can print something like `'starting training with model:\n{}'`

**Review by Canoming — COMMENTED** [review]
I think the issue is that, we are still not clear about what should be shown if we want a `verbose training

**Comment by Canoming** [comment]
We may also want to include the process id to the printed result for multi-processing

**Comment by cfenglv** [comment]
The added tests for back minimize also comes in this pr

**Comment by cfenglv** [comment]
> We may also want to include the process id to the printed result for multi-processing

Process id included.
